src/Microsoft.Identity.Web/TokenCacheProviders/Distributed/MsalDistributedTokenCacheAdapter.cs

```diff
@@ -10,210 +10,210 @@
+		/// </summary>
+		private readonly MsalDistributedTokenCacheAdapterOptions _distributedCacheOptions;
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="MsalDistributedTokenCacheAdapter"/> class.
+		/// </summary>
+		/// <param name="distributedCache">Distributed cache instance to use.</param>
+		/// <param name="distributedCacheOptions">Options for the token cache.</param>
+		/// <param name="logger">MsalDistributedTokenCacheAdapter logger.</param>
+		public MsalDistributedTokenCacheAdapter(
+											IDistributedCache distributedCache,
+											IOptions<MsalDistributedTokenCacheAdapterOptions> distributedCacheOptions,
+											ILogger<MsalDistributedTokenCacheAdapter> logger)
+		{
+			if (distributedCacheOptions == null)
+			{
+				throw new ArgumentNullException(nameof(distributedCacheOptions));
+			}
+
+			_distributedCache = distributedCache;
+			_distributedCacheOptions = distributedCacheOptions.Value;
+			_memoryCache = new MemoryCache(_distributedCacheOptions.L1CacheOptions ?? new MemoryCacheOptions { SizeLimit = 500 * 1024 * 1024 });
+			_logger = logger;
+
+			if (_distributedCacheOptions.AbsoluteExpirationRelativeToNow != null)
+			{
+				if (_distributedCacheOptions.L1ExpirationTimeRatio <= 0 || _distributedCacheOptions.L1ExpirationTimeRatio > 1)
+				{
+					throw new ArgumentOutOfRangeException(nameof(_distributedCacheOptions.L1ExpirationTimeRatio), "L1ExpirationTimeRatio must be greater than 0, less than 1. ");
+				}
+
+				_expirationTime = TimeSpan.FromMilliseconds(_distributedCacheOptions.AbsoluteExpirationRelativeToNow.Value.TotalMilliseconds * _distributedCacheOptions.L1ExpirationTimeRatio);
+			}
+		}
+
+		/// <summary>
+		/// Removes a specific token cache, described by its cache key
+		/// from the distributed cache.
+		/// </summary>
+		/// <param name="cacheKey">Key of the cache to remove.</param>
+		/// <returns>A <see cref="Task"/> that completes when key removal has completed.</returns>
+		protected override async Task RemoveKeyAsync(string cacheKey)
+		{
+			_memoryCache.Remove(cacheKey);
+			_logger.LogDebug($"[MsIdWeb] MemoryCache: Removed cacheKey {cacheKey}. ");
+
+			await L2OperationWithRetryOnFailureAsync(
+				"Removed",
+				(cacheKey) => _distributedCache.RemoveAsync(cacheKey),
+				cacheKey).ConfigureAwait(false);
+		}
+
+		/// <summary>
+		/// Read a specific token cache, described by its cache key, from the
+		/// distributed cache.
+		/// </summary>
+		/// <param name="cacheKey">Key of the cache item to retrieve.</param>
+		/// <returns>Read blob representing a token cache for the cache key
+		/// (account or app).</returns>
+		protected override async Task<byte[]> ReadCacheBytesAsync(string cacheKey)
+		{
+			// check memory cache first
+			byte[]? result = (byte[])_memoryCache.Get(cacheKey);
+			_logger.LogDebug($"[MsIdWeb] MemoryCache: Read {cacheKey} cache size: {result?.Length ?? 0}. ");
+
+			if (result == null)
+			{
+				var measure = await Task.Run(async () =>
```

Comment:
Using Task.Run is usually something that shouldn't be done in an already async function for perf/context switching reasons. It looks like it is only done to measure things. I suggest removing task.run and measuring with StopWatch.StartNew.